File: 28-03-2024_Length_of_Longest_Subarray_with_atmost_K_frequency.cpp
class Solution {
public:
    int maxSubarrayLength(vector<int>& nums, int k) {
        unordered_map<int, int> mp;
        int n = nums.size(), i = 0, j = 0, ans = 0, flag = 0;
        while(j < n){
            if(++mp[nums[j]] > k) flag = 1;
            while(flag and i <= j){
                ans = max(ans, j - i);
                --mp[nums[i++]];
                if(mp[nums[j]] == k) flag = 0;
            }
            j++;
        }
        ans = max(ans, j - i);
        return ans;       
    }
};

File: 28-03-2024_Length_of_Longest_Subarray_with_atmost_K_frequency.py
class Solution:
    def maxSubarrayLength(self, nums: List[int], k: int) -> int:
        mp = {}
        n = len(nums)
        i, j, ans, flag = 0, 0, 0, 0
        while j < n:
            mp[nums[j]] = mp.get(nums[j], 0) + 1
            if mp[nums[j]] > k:
                flag = 1
            while flag == 1 and i <= j:
                ans = max(ans, j - i)
                mp[nums[i]] -= 1
                if mp[nums[j]] == k:
                    flag = 0
                i += 1
            j += 1
        ans = max(ans, j - i)
        return ans
